src/dbnode/storage/series/buffer.go

```diff
@@ -181,287 +161,281 @@ func (b *dbBuffer) Write(
-			return
-		}
-
-		if !bucket.canRead() {
-			return
-		}
+	stream, err := block.Stream(ctx)
+	if err != nil {
+		return FlushOutcomeErr, err
+	}
 
-		if !blockStart.Equal(bucket.start) {
-			return
-		}
+	segment, err := stream.Segment()
+	if err != nil {
+		return FlushOutcomeErr, err
+	}
 
-		// We need to merge all the bootstrapped blocks / encoders into a single stream for
-		// the sake of being able to persist it to disk as a single encoded stream.
-		_, err = bucket.merge()
-		if err != nil {
-			return
-		}
+	checksum, err := block.Checksum()
+	if err != nil {
+		return FlushOutcomeErr, err
+	}
 
-		// This operation is safe because all of the underlying resources will respect the
-		// lifecycle of the context in one way or another. The "bootstrapped blocks" that
-		// we stream from will mark their internal context as dependent on that of the passed
-		// context, and the Encoder's that we stream from actually perform a data copy and
-		// don't share a reference.
-		streams := bucket.streams(ctx)
-		if len(streams) != 1 {
-			// Should never happen as the call to merge above should result in only a single
-			// stream being present.
-			err = errMoreThanOneStreamAfterMerge
-			return
-		}
+	err = persistFn(id, tags, segment, checksum)
+	if err != nil {
+		return FlushOutcomeErr, err
+	}
 
-		// Direct indexing is safe because canRead guarantees us at least one stream
-		res = streams[0]
-	})
+	// Don't need to check error here because a non-nil block means that a
+	// writable bucket exists
+	bucket, _ := buckets.writableBucket(WarmWrite)
+	bucket.version = version
 
-	return res, err
+	return FlushOutcomeFlushedToDisk, nil
 }
 
 func (b *dbBuffer) ReadEncoded(ctx context.Context, start, end time.Time) [][]xio.BlockReader {
 	// TODO(r): pool these results arrays
 	var res [][]xio.BlockReader
-	b.forEachBucketAsc(func(bucket *dbBufferBucket) {
-		if !bucket.canRead() {
-			return
-		}
-		if !start.Before(bucket.start.Add(b.blockSize)) {
-			return
-		}
-		if !bucket.start.Before(end) {
-			return
+
+	keys := b.sortedBucketsKeys(true)
+	for _, key := range keys {
+		buckets, exists := b.bucketsAt(key.ToTime())
+		if !exists || !buckets.start.Before(end) ||
+			!start.Before(buckets.start.Add(b.blockSize)) {
+			continue
 		}
 
-		res = append(res, bucket.streams(ctx))
+		res = append(res, buckets.streams(ctx, WarmWrite))
+		res = append(res, buckets.streams(ctx, ColdWrite))
 
 		// NB(r): Store the last read time, should not set this when
 		// calling FetchBlocks as a read is differentiated from
 		// a FetchBlocks call. One is initiated by an external
 		// entity and the other is used for streaming blocks between
 		// the storage nodes. This distinction is important as this
 		// data is important for use with understanding access patterns, etc.
-		bucket.setLastRead(b.nowFn())
-	})
+		buckets.setLastRead(b.nowFn())
+	}
 
 	return res
 }
 
+func (b *dbBuffer) sortedBucketsKeys(ascending bool) []xtime.UnixNano {
```

Comment:
So I think the reason this used to accept a function was to avoid allocating a slice that definetly escapes. Is there a reason we're switching to doing it this way now?